xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -259,35 +258,87 @@ public void run() {
         }
         ResourceSubscriber<T> subscriber =
             (ResourceSubscriber<T>) resourceSubscribers.get(type).get(resourceName);
+
         if (subscriber == null) {
           logger.log(XdsLogLevel.INFO, "Subscribe {0} resource {1}", type, resourceName);
           subscriber = new ResourceSubscriber<>(type, resourceName);
           resourceSubscribers.get(type).put(resourceName, subscriber);
-          if (subscriber.controlPlaneClient != null) {
-            subscriber.controlPlaneClient.adjustResourceSubscription(type);
+          subscriber.addWatcher(watcher, watcherExecutor);
+          boolean hadErrorBeforeCpc = subscriber.errorDescription != null;
+
+          ControlPlaneClient cpcToUse = subscriber.loadControlPlane();
+
+          if (subscriber.errorDescription != null) {
+            // Notify the watchers of the error if we didn't do it already in addWatcher
+            if (!hadErrorBeforeCpc) {
+              subscriber.onError(
+                  Status.UNAVAILABLE.withDescription(subscriber.errorDescription), null);
+            }
+            return;
           }
+
+          doFallbackIfNecessary(cpcToUse, subscriber.authority);
```

Comment:
This method re-discovers the code flow taken by `getOrCreateControlPlaneClient()`. Can that be integrated into one method to avoid the error-prone handling? Every call to `getOrCreateControlPlaneClient()` needs to add the fallback if necessary, and the only other call to `getOrCreateControlPlaneClient()` just calls `fallBackToCpc()` unconditionally. It would be better to create and update the state together in one method so we know the state remains consistent.